### PR TITLE
Add which() to CommandExecutor for finding commands

### PR DESCRIPTION
**Pull request by mfurtak**
Add a cross-platform way of finding an executable in the $PATH. Respects the $PATHEXT, which lists valid file extensions for executables on Windows.

### CODE REVIEW

File: lib/fastlane_core/command_executor.rb [hunk 1]
```diff
@@ -2,6 +2,25 @@ module FastlaneCore
...
   class CommandExecutor
     class << self
+      # Cross-platform way of finding an executable in the $PATH. Respects the $PATHEXT, which lists
+      # valid file extensions for executables on Windows.
+      #
+      #    which('ruby') #=> /usr/bin/ruby
+      #
+      # Derived from http://stackoverflow.com/a/5471032/3005
+      def which(cmd)
+        exts = ENV['PATHEXT'] ? ENV['PATHEXT'].split(';') : ['']
```
Comment (mfurtak) [inline lib/fastlane_core/command_executor.rb]:
```
 exts = (ENV['PATHTEXT'] || []).split(';')
```

Comment (mfurtak) [inline lib/fastlane_core/command_executor.rb]: `executable_exts`?

File: spec/command_executor_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+          temp_dir = File.dirname(f)
+          temp_cmd = File.basename(f, '.exe')
+
+          with_env_values('PATH' => temp_dir, 'PATHEXT' => '.exe') do
+            expect(FastlaneCore::CommandExecutor.which(temp_cmd)).to eq(f.path)
+          end
+        end
+      end
+
+      it "does not find commands with unknown extensions which are on the PATH" do
```
Comment (mfurtak) [inline spec/command_executor_spec.rb]: great tests! :+1:

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -10,3 +10,19 @@ module SpecHelper
 end
 
 WebMock.disable_net_connect!(allow: 'coveralls.io')
+
+# Executes the provided block after adjusting the ENV to have the
+# provided keys and values set as defined in hash. After the block
+# completes, restores the ENV to its previous state.
+def with_env_values(hash)
+  old_vals = ENV.select { |k, v| hash.include?(k) }
```
Comment (mfurtak) [inline spec/spec_helper.rb]: just make a copy of the whole has and restore it?

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -10,3 +10,19 @@ module SpecHelper
...
+  hash.each do |k, v|
+    ENV[k] = hash[k]
+  end
+  yield
```
Comment (mfurtak) [inline spec/spec_helper.rb]: smart!

**Comment by mfurtak** [comment]
These comments were SamP giving me feedback from my machine

File: spec/spec_helper.rb (hunk 3 above)
Comment (mfurtak) [inline spec/spec_helper.rb]: Sadly `ENV` has some of the methods of a hash, but is actually an `Object` :frowning:

File: lib/fastlane_core/command_executor.rb (hunk 1 above)
Comment (KrauseFx) [inline lib/fastlane_core/command_executor.rb]: Yeah, using the `||` is nicer

Comment (mfurtak) [inline lib/fastlane_core/command_executor.rb]: Sadly this was not equivalent :frowning:

**Comment by mfurtak** [comment]
@KrauseFx The plan is to use this immediately in chiizu. Please let me know what you think!

**Comment by KrauseFx** [comment]
Looks good to me, great work @mfurtak.

**Comment by KrauseFx** [comment]
Nice `danger` :rocket:

**Pull request merged**